File: classlink-oneroster-oauth1/node/OneRoster.js
const CryptoJs = require('crypto-js');
const request = require('request');

/**
 * Key and secret for the request
 * @constructor
 * @param {string} clientId
 * @param {string} clientSecret 
 */
function OneRoster(clientId, clientSecret) {
    this.clientId = clientId;
    this.clientSecret = clientSecret;
}

/**
 * Makes a GET request to the given url with the stored key and secret
 * @param {*} url  The url of the request
 * @param {*} callback   The callback to be executed after the requesta
 */
OneRoster.prototype.makeRosterRequest = function(url, callback) {

    // Generate timestamp and nonce
    let timestamp = time().toString();
    let nonce = generateNonce(timestamp.length);
    
    // Definte oauth params
    let oauth = {
        'oauth_consumer_key': this.clientId,
        'oauth_signature_method': 'HMAC-SHA256',
        'oauth_timestamp': timestamp,
        'oauth_nonce': nonce
    };

    // Create object with all the params from the url mixed with oauth params
    let urlPieces = url.split("?");
    let params;
    let query;
    if (urlPieces.length == 2) {
        let urlParams = paramsToObject(urlPieces[1]);
        query = Object.assign({}, urlParams);
        params = Object.assign(urlParams, oauth);
    } else {
        query = {};
        params = oauth;
    }
    
    // Generate oauth signature
    let baseInfo = buildBaseString(urlPieces[0], 'GET', params);
    let compositeKey = encodeURIComponent(this.clientSecret) + "&";
    let oauthSignature = CryptoJs.enc.Base64.stringify(CryptoJs.HmacSHA256(baseInfo, compositeKey));
    oauth.oauth_signature = oauthSignature;

    // Generate header and make request
    let authHeader = buildAuthorizationHeader(oauth);

    makeRequest(authHeader, urlPieces[0], query, callback);
};

/**
 * Builds an authorization header for the request
 * @param {object} oauthInfo The oauth header params
 * @returns Returns the authorization header for the request as a string
 */
let buildAuthorizationHeader = function(oauthInfo) {
    let result = 'OAuth ';
    values = [];
    for (let key in oauthInfo) {
        values.push(key + '="' + encodeURIComponent(oauthInfo[key]) + '"');
    }
    result += values.join(",");
    return result;
};

/**
 * Generates a random nonce
 * @param {number} length The length of the non
 * @returns The nonce as a string
 */
let generateNonce = function(length) {
    let text = "";
    let possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
  
    for (let i = 0; i < length; i++)
      text += possible.charAt(Math.floor(Math.random() * possible.length));
  
    return text;
  };

/**
 * Converts a string of url parameters into an object
 * @param {string} urlParams A string with the parameters from the URL
 * @returns An object with the parameters
 */
let paramsToObject = function(urlParams) {
    let params = urlParams.trim().split('&');
    let result = {};
    for(let i = 0; i < params.length; i++) {
        if (typeof(params[i]) == undefined) {
            break;
        }

        let value = decodeURIComponent(params[i]);
        let paramParts = value.split("=");
        if (paramParts.length == 2) {
            result[paramParts[0]] = paramParts[1];
        } else {
            result.filter = value.substr(7);
        }
    }
    return result;
};

/**
 * Sorts the keys of an object
 * @param {object} o the object to be sorted
 * @returns The object with its keys sorted
 */
let sortObject = function(o) {
    return Object.keys(o).sort().reduce((r, k) => (r[k] = o[k], r), {});
};

let buildBaseString = function(baseURI, method, params) {
    let r = [];
    params = sortObject(params);
    for (let key in params) {
        let value = params[key];
        value = encodeURIComponent(value);
        value = value.replace(/'/g, "%27");
        r.push(key + "=" + value);
    }

    return method + "&" + encodeURIComponent(baseURI) + "&" + encodeURIComponent(r.join("&"));

};

/**
 * Generates a timestamp
 * @returns a timestamp
 */
let time = function() {
    let timestamp = Math.floor(new Date().getTime() / 1000);
    return timestamp;
};

/**
 * Makes a GET request to the url with the given parameters and the generated authoriztaion header
 * @param {string} authHeader The authoriation header for the request
 * @param {string} url The base url for the request
 * @param {ojbect} urlParams The url params for the request
 * @param {function} callback The callback to be executed after the request
 */
let makeRequest = function(authHeader, url, urlParams, callback) {

    let options = { method: 'GET',
        url: url,
        qs: urlParams,
        headers: 
        { 'Authorization':  authHeader}
    };

    return request(options, function (error, response, body) {
        if (error) {
            callback(error, 0, body);
        } else {
            callback(error, response.statusCode, body);
        }
    });
};

module.exports = OneRoster;

File: classlink-oneroster-oauth1/dotnetframework/OneRoster/OneRoster/OneRoster.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;
using System.Web;

namespace ClassLink.OneRoster
{
    public class OneRoster
    {
        private readonly string _clientId;
        private readonly string _clientSecret;

        public OneRoster(string clientId, string clientSecret)
        {
            _clientId = clientId;
            _clientSecret = clientSecret;
        }

        /// <summary>
        /// Make a GET request to the given url with the stored key and secret
        /// </summary>
        /// <param name="url">The url for the request, with params included</param>
        /// <returns>A dictionary of the status_code and response</returns>
        public OneRosterResponse MakeRosterRequest(string url)
        {
            // Generate timestamp and nonce
            string timestamp = ((int)(DateTime.UtcNow - new DateTime(1970, 1, 1)).TotalSeconds).ToString();
            string nonce = GenerateNonce(timestamp.Length);

            // Definte oauth params
            Dictionary<string, string> oauth = new Dictionary<string, string>
            {
                {"oauth_consumer_key", _clientId},
                {"oauth_signature_method", "HMAC-SHA256"},
                {"oauth_timestamp", timestamp},
                {"oauth_nonce", nonce}
            };

            // Combine url params and oauth params into a sorted dictionary
            string[] urlPieces = url.Split('?');
            SortedDictionary<string, string> allParams;
            if (urlPieces.Length == 2)
            {
                Dictionary<string, string> urlParams = ParamsToDic(urlPieces[1]);
                allParams = SortAllParams(urlParams, oauth);
            } else
            {
                allParams = SortAllParams(oauth, new Dictionary<string, string>());
            }


            // Generate the signature
            string baseInfo = BuildBaseString(urlPieces[0], "GET", allParams);
            string compositeKey = UrlEncodeUpperCase(_clientSecret) + "&";
            string authSignature = GenerateSig(baseInfo, compositeKey);
            oauth.Add("oauth_signature", authSignature);

            // Generate the header and make the request 
            string authHeader = BuildAuthHeader(oauth);
            return MakeGetRequest(authHeader, url);
        }

        /// <summary>
        /// Create a random string for the nonce
        /// </summary>
        /// <param name="len">The length of the nonce</param>
        /// <returns></returns>
        private string GenerateNonce(int len)
        {
            const string allowedChars = "ABCDEFGHJKLMNOPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz0123456789";
            char[] chars = new char[len];

            Random rd = new Random();

            for (int i = 0; i < len; i++)
            {
                chars[i] = allowedChars[rd.Next(0, allowedChars.Length)];
            }

            return new string(chars);
        }

        /// <summary>
        /// Make the request to the given url with generated header
        /// </summary>
        /// <param name="authHeader">The generated auth header</param>
        /// <param name="url">The url for the request</param>
        /// <returns>A dictionary with the status_code and response</returns>
        private OneRosterResponse MakeGetRequest(string authHeader, string url)
        {
            try
            {
                HttpWebRequest webRequest = (HttpWebRequest)HttpWebRequest.Create(url);
                if (webRequest != null)
                {
                    webRequest.Method = "GET";
                    webRequest.Timeout = 25000;
                    webRequest.Headers.Add("Authorization", authHeader);

                    HttpWebResponse webResponse = (HttpWebResponse)webRequest.GetResponse();

                    using (Stream s = webResponse.GetResponseStream())
                    {
                        using (StreamReader sr = new StreamReader(s))
                        {
                            var response = sr.ReadToEnd();
                            return new OneRosterResponse((int)webResponse.StatusCode, response);
                        }
                    }
                }
            } catch (WebException webException)
            {
                try
                {
                    using (var sr = new StreamReader(webException.Response.GetResponseStream()
                                                     ?? throw new Exception()))
                    {
                        return new OneRosterResponse((int)webException.Status, sr.ReadToEnd());
                    }
                } catch (Exception)
                {
                    return new OneRosterResponse(0, "An error occurred, check your URL");
                }

            } catch (Exception)
            {
                return new OneRosterResponse(0, "An error occurred, check your URL");
            }

            return null;
        }

        /// <summary>
        /// Generates the auth header from the oauth params
        /// </summary>
        /// <param name="oauth">A dictionary of oauth params</param>
        /// <returns>The auth header</returns>
        private string BuildAuthHeader(Dictionary<string, string> oauth)
        {
            string result = "OAuth ";
            string[] values = new string[oauth.Count];
            int i = 0;
            foreach (var key in oauth.Keys)
            {
                values[i] = $"{key}=\"{UrlEncodeUpperCase(oauth[key])}\"";
                i++;
            }
            result += string.Join(",", values);
            return result;
        }

        /// <summary>
        /// Generate the oauth signature from the base string and key
        /// </summary>
        /// <param name="baseString">The base string generated from method, url, and params</param>
        /// <param name="key">The key created from ClientSecret</param>
        /// <returns>The generated oauth signature</returns>
        private string GenerateSig(string baseString, string key)
        {
            Encoding encoding = new ASCIIEncoding();
            byte[] byteKey = encoding.GetBytes(key);
            byte[] stringBytes = encoding.GetBytes(baseString);

            string result = string.Empty;
            using (HMACSHA256 sha256 = new HMACSHA256(byteKey))
            {
                var hashed = sha256.ComputeHash(stringBytes);
                result = Convert.ToBase64String(hashed);
            }


            return result;

        }

        /// <summary>
        /// Builds the base string for the generation of the oauth signature
        /// </summary>
        /// <param name="baseUrl">The base url without params</param>
        /// <param name="method">The request's HTTP method</param>
        /// <param name="allParams">The url params and oauth params</param>
        /// <returns>The base string</returns>
        private string BuildBaseString(string baseUrl, string method, SortedDictionary<string, string> allParams)
        {
            string[] result = new string[allParams.Count];
            int i = 0;
            foreach (var key in allParams.Keys)
            {
                result[i] = $"{key}={UrlEncodeUpperCase(allParams[key])}";
                i++;

            }

            return method + "&" + UrlEncodeUpperCase(baseUrl) + "&" +
                   UrlEncodeUpperCase(string.Join("&", result));
        }

        /// <summary>
        /// URL encodes the string
        /// </summary>
        /// <param name="value">The string to encode</param>
        /// <returns>The encoded string</returns>
        private string UrlEncodeUpperCase(string value)
        {
            value = HttpUtility.UrlEncode(value);
            value = value.Replace("+", "%20");
            return Regex.Replace(value, "(%[0-9a-f][0-9a-f])", c => c.Value.ToUpper());
        }

        /// <summary>
        /// Combines all the params into one sorted dictionary
        /// </summary>
        /// <param name="urlParams">The params from the url</param>
        /// <param name="oauth">The oauth params</param>
        /// <returns>A sorted dictionary of all the params</returns>
        private SortedDictionary<string, string> SortAllParams(Dictionary<string, string> urlParams,
            Dictionary<string, string> oauth)
        {
            SortedDictionary<string, string> result = new SortedDictionary<string, string>();
            foreach (var key in urlParams.Keys)
            {
                result.Add(key, urlParams[key]);
            }

            foreach (var key in oauth.Keys)
            {
                result.Add(key, oauth[key]);
            }

            return result;
        }

        /// <summary>
        /// Converts the params from the url to a dictionary
        /// </summary>
        /// <param name="urlPiece">The params in the url</param>
        /// <returns>A dictionary of the params</returns>
        private Dictionary<string, string> ParamsToDic(string urlPiece)
        {
            string[] theParams = urlPiece.Split('&');
            Dictionary<string, string> result = new Dictionary<string, string>();
            foreach (var value in theParams)
            {
                string decodedVal = HttpUtility.UrlDecode(value);
                string[] split = decodedVal.Split('=');
                if (split.Length == 2)
                {
                    result.Add(split[0], split[1]);
                } else
                {
                    result.Add("filter", decodedVal.Substring(7));
                }
            }

            return result;
        }
    }
}